ENG-2461: Fix for cert pinning issue when requesting multiple APIs

balu-eligible opened the pull request:
JIRA: https://eligible.atlassian.net/browse/ENG-2461

Customer is using multiple APIs along with eligibleAPI. If on connecting to Eligible and then other APIs are failing with error "Authentication Error:Connection is closed." So that other APIs are not working if they are running EligibleAPI.
If on running EligibleAPI seperately they are working.

Issue:  We are using ServicePointManager for cert pinning.
ServicePointManager has only static members.So, it is global for AppDomain means it is global for the whole .NET application. 
Solution: I added a IsEligibleRequest property to tweak this default functionality. Before executing any eligible request we set IsEligibleRequest=true, once done we set it back to false. So, with this we can assure that other HTTPs calls will not be effected with our cert pinning because they have by default IsEligibleRequest=false(we always return true means we are ignoring other HTTPS calls). 

### Checklists
**Development and testing:**
- [x] All tests passed on Local environment

**Pre-release and code review:**
- [x] CI build is green
- [x] Changes have been reviewed by at least one Senior Engineer
- [x] JIRA issue has a link to this pull request

kovyrin commented:
The fact that we need to manually manage that flag from all over the code smells bad to me... What if we were to look at the domain of the certificate and use that as our cue that the request is to Eligible?

Something like:
```ruby
is_eligible = Eligible.whitelisted_domain?(certificate.domain)
```

balu-eligible commented:
@kovyrin will do that and let you know once done....

balu-eligible commented:
@kovyrin PR updated. Please review when you get some time.

kovyrin commented:
Oh, this looks so much prettier. @abedra, any concerns or comments about the new approach?

abedra commented:
This looks like a nice cleanup. I noticed some tests got deleted. Are those no longer needed or did we forget to add them back in/correct them?

balu-eligible commented:
@abedra I deleted one task. that is actually no use. 

I added this test case some some time back, was trying to test the failure case of wrong certificate fingerprint, but it is completely no sense because we are using fingerprints that are present in resource file, and it wont consider the fingerprint that added in test case. I found this while debugging so, deleting from suite.

abedra commented:
Ok. Do we still have a test that demonstrates the behavior of the code when an invalid fingerprint is encountered?

balu-eligible commented:
@abedra We can do that but as we discussed in the initial days of implementation we made fingerprint property as private so, outside people cant add fingerprints.. If i want to test this i need to made it as public, so it breaks what we discussed. Any suggessions?

abedra commented:
This test serves two purposes. It ensures that if the fingerprint for the handshake is ever incorrect, our code will do what we expect, and to let people that are using our code see what will happen if that is ever encountered.

On another note, we should still have the "escape hatch" that allows the user to specify a fingerprint in case of emergency. We need this because we pin to the leaf and if there ever needs to be rapid rotation of our current and backup certs we don't want our users to stop working.

balu-eligible commented:
@abedra We have a method **AddFingerPrint** in case of emergency.
I will work on the failure test case will back soon

abedra commented on Eligible.NET/src/EligibleService.Common/RequestExecute.cs:
```diff
@@ -122,7 +122,14 @@ public bool CertificateValidation(object sender, X509Certificate certificate, X5
                 return false;
 
-            if (certificate.Subject.Contains(EligibleResources.WhiteListedDomain.Trim()))
+            bool is_white_listed_domain = false;
```
What's the purpose of this new code?

balu-eligible replied:
I am trying to add property that can be used to add whitelist domains from outside app, in case if there if there are any change in our certificate domain. This also helps me to add a new domain from out side and can be able to test failures case of wrong fingerptint.

balu-eligible commented:
@abedra Please review the update PR, when you get some time.

abedra commented:
Ok lgtm :shipit:

balu-eligible commented:
Thank you @abedra . merging the changes now
